[DONE] [BUG] Some issue with CHANGELOG (recurrent) wherein it has a double blank line, thus failing CI on first run

Log:
```
⏺ Bash(IMPLEMENT_TMPDIR=/Users/zhupanov/.cache/larch/sessions/claude-implement-larch5-26JKm9
      LARCH_TOKEN_SESSION_ID=$("/Users/zhupanov/.claude/plugins/cache/larch-loca…)
  ⎿  ⏳ CI: waiting..
     → Action: evaluate_failure (31s, 2 polls)
     ACTION=evaluate_failure
     … +6 lines (ctrl+o to expand)
  ⎿  (timeout 10m)

⏺ Bash(/Users/zhupanov/.claude/plugins/cache/larch-local/larch/17.0.26/scripts/gh-run-logs.sh --run-id 25583918291 --repo character-ai/larch 2>&1 | tail -80)
  ⎿  run 25583918291 is still in progress; logs will be available when it is complete

⏺ Bash(gh run view 25583918291 --repo character-ai/larch --json conclusion,status,name,jobs --jq '{conclusion, status, name, failed_jobs: [.jobs[] | select(.conclusion…)
  ⎿  {
       "conclusion": "",
       "failed_jobs": [
     … +18 lines (ctrl+o to expand)

⏺ Bash(gh run view --job 75108558406 --repo character-ai/larch --log 2>&1 | tail -60)
  ⎿  run 25583918291 is still in progress; logs will be available when it is complete

⏺ Bash(gh run view --job 75108558406 --repo character-ai/larch 2>&1 | tail -50)
  ⎿  * sergey-zhupanov/relevant-checks-script-first CI character-ai/larch#1575 · 25583918291
     Triggered via pull_request about 1 minute ago
     … +23 lines (ctrl+o to expand)

⏺ Bash(gh run view --log-failed --job=75108558406 --repo character-ai/larch 2>&1 | tail -60)
  ⎿  run 25583918291 is still in progress; logs will be available when it is complete

⏺ Bash(/Users/zhupanov/.claude/plugins/cache/larch-local/larch/17.0.26/scripts/sleep-seconds.sh 60
      gh run view 25583918291 --repo character-ai/larch --json conclusion,…)
  ⎿  {
       "conclusion": "failure",
       "status": "completed"
     }
  ⎿  (timeout 5m)

⏺ Bash(gh run view --log-failed --job=75108558406 --repo character-ai/larch 2>&1 | grep -E '(error|fail|Detected|FAIL|ERROR)' | head -40)
  ⎿  lint       Run pre-commit linters  2026-05-08T23:08:33.5511417Z ##[error]Process completed with exit code 2.
  ⎿  (timeout 5m)

⏺ Bash(gh run view --log-failed --job=75108558406 --repo character-ai/larch 2>&1 | tail -50)
  ⎿  lint       Run pre-commit linters  2026-05-08T23:08:02.9539993Z ##[group]Run make lint-only
     lint       Run pre-commit linters  2026-05-08T23:08:02.9540325Z ^[[36;1mmake lint-only^[[0m
     lint       Run pre-commit linters  2026-05-08T23:08:02.9562080Z shell: /usr/bin/bash -e {0}
     … +24 lines (ctrl+o to expand)

⏺ CHANGELOG has double-blank line. Fix.

  Read 1 file (ctrl+o to expand)
```

Claude's analysis:
    Likely culprit: the helper inside implement-finalize.sh (which delegates to a script like scripts/changelog-insert.sh or similar — I don't recall the exact name from context) inserts ## [NEW_VERSION] - YYYY-MM-DD\n\n### Changed\n\n-
    bullet\n\n before the next existing ## [X.Y.Z] header, but the next header is itself preceded by its own \n separator, producing \n\n + \n## [...] = three newlines = a double blank line.

    The fix would belong in that insertion script (collapse the trailing newlines of the new section to exactly one blank line before the next header), and would need a regression test in the changelog-amend harness. I did not file an issue
    or fix that — I just hand-edited the rendered output. Worth a follow-up OOS issue.

**PR**: https://github.com/character-ai/larch/pull/1582

--- a/.claude-plugin/plugin.json
+++ b/.claude-plugin/plugin.json
@@ -1,6 +1,6 @@
 {
   "name": "larch",
-  "version": "18.0.9",
+  "version": "18.0.10",
   "description": "Multi-agent workflow automation for Claude Code: design, code review, implementation, and PR automation. Code review runs a 3-reviewer specialist panel (2 Cursor specialists + 1 Codex generic) — extended to 4 with an optional Gemini reviewer when the Gemini CLI is installed and healthy — with 3-voter adjudication; plan review runs a 3-reviewer panel (Claude + Codex + Cursor); design sketches run a 9-agent diverge-then-converge phase in regular mode (1 Claude + 4 Cursor + 4 Codex) or 3 in quick mode.",
   "author": {
     "name": "Sergey Zhupanov",

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [18.0.10] - 2026-05-08
+
+### Changed
+
+- Fix recurrent double blank line after CHANGELOG entry by fixing separator logic in write_changelog_entry and changelog_categories_to_markdown
+
 ## [18.0.9] - 2026-05-08
 
 ### Added

--- a/scripts/implement-finalize.sh
+++ b/scripts/implement-finalize.sh
@@ -502,6 +502,7 @@ write_changelog_entry() {
             skipping = 0
             in_unreleased = 0
             match_count = 0
+            entry_from_version_match = 0
         }
         FNR == NR {
             if (/^## \[Unreleased\]/) has_unreleased = 1
@@ -516,11 +517,13 @@ write_changelog_entry() {
             if (!inserted) {
                 for (i = 1; i <= en; i++) print e[i]
                 inserted = 1
+                entry_from_version_match = 1
             }
             skipping = 1
             next
         }
         skipping && /^## \[/ {
+            if (entry_from_version_match) print ""
             skipping = 0
         }
         skipping {

--- a/scripts/implement-finalize.sh
+++ b/scripts/implement-finalize.sh
@@ -435,13 +435,13 @@ changelog_categories_to_markdown() {
     for category in Added Changed Fixed Removed Security; do
         file="$dir/$category"
         [ -s "$file" ] || continue
+        [ "$had_any" = "true" ] && printf '\n' >> "$out"
         had_any=true
         printf '### %s\n\n' "$category" >> "$out"
         while IFS= read -r title || [ -n "$title" ]; do
             [ -n "$title" ] || continue
             printf -- '- %s\n' "$title" >> "$out"
         done < "$file"
-        printf '\n' >> "$out"
     done
     [ "$had_any" = "true" ]
 }
@@ -556,6 +556,7 @@ write_changelog_entry() {
         }
         !inserted && /^## \[/ {
             for (i = 1; i <= en; i++) print e[i]
+            print ""
             inserted = 1
         }
         { print }

--- a/scripts/test-implement-finalize.sh
+++ b/scripts/test-implement-finalize.sh
@@ -44,6 +44,12 @@ assert_file_contains() {
     assert_contains "$needle" "$content" "$label"
 }
 
+assert_file_not_contains() {
+    local needle=$1 path=$2 label=$3 content
+    content=$(cat "$path" 2>/dev/null || true)
+    assert_not_contains "$needle" "$content" "$label"
+}
+
 assert_rc() {
     local actual=$1 expected=$2 label=$3
     if [ "$actual" -eq "$expected" ]; then
@@ -728,6 +734,9 @@ assert_file_contains "--repo" "$SANDBOX/refresh-anchor-argv.txt" "postbump: anch
 assert_file_contains "IMPLEMENT_TMPDIR=$SANDBOX/tmp" "$SANDBOX/refresh-anchor-env.txt" "postbump: exports IMPLEMENT_TMPDIR to children"
 assert_file_contains "## [17.0.4] -" "$SANDBOX/repo/CHANGELOG.md" "postbump: changelog contains new version"
 assert_file_contains "### Changed" "$SANDBOX/repo/CHANGELOG.md" "postbump: flat bullets default to Changed"
+awk 'prev_blank && /^$/ { print "DOUBLE_BLANK"; exit } /^$/ { prev_blank=1; next } { prev_blank=0 }' \
+    "$SANDBOX/repo/CHANGELOG.md" > "$SANDBOX/blank-check.txt"
+assert_file_not_contains "DOUBLE_BLANK" "$SANDBOX/blank-check.txt" "postbump: changelog has no consecutive blank lines"
 
 # FINDING_1 regression: Unreleased section bullets must be preserved, not consumed.
 cat > "$SANDBOX/repo/CHANGELOG.md" <<'CHANGELOG_UNRELEASED'

--- a/scripts/test-implement-finalize.sh
+++ b/scripts/test-implement-finalize.sh
@@ -824,6 +824,34 @@ assert_file_contains "### Fixed" "$SANDBOX/repo/CHANGELOG.md" "postbump: fallbac
 assert_file_contains "### Added" "$SANDBOX/repo/CHANGELOG.md" "postbump: fallback categorized Added"
 assert_file_contains "### Changed" "$SANDBOX/repo/CHANGELOG.md" "postbump: fallback bare bullet defaults Changed"
 
+# Regression: when target version header already exists (replacement path), no double blank before next header.
+cat > "$SANDBOX/repo/CHANGELOG.md" <<'CHANGELOG_REPLACE'
+# Changelog
+
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [17.0.4] - 2026-05-01
+
+### Fixed
+
+- old content to be replaced.
+
+## [17.0.3] - 2026-04-30
+
+### Changed
+
+- Prior change.
+CHANGELOG_REPLACE
+write_postbump_state "$POSTBUMP_STATE"
+OUT=$(run_subject postbump --state-file "$POSTBUMP_STATE" --implement-tmpdir "$SANDBOX/tmp")
+assert_contains "CHANGELOG_STATUS=updated" "$OUT" "postbump: replacement path updates changelog"
+awk 'prev_blank && /^$/ { print "DOUBLE_BLANK"; exit } /^$/ { prev_blank=1; next } { prev_blank=0 }' \
+    "$SANDBOX/repo/CHANGELOG.md" > "$SANDBOX/blank-check-replace.txt"
+assert_file_not_contains "DOUBLE_BLANK" "$SANDBOX/blank-check-replace.txt" "postbump: replacement path has no consecutive blank lines"
+
 cp "$SANDBOX/original-CHANGELOG.md" "$SANDBOX/repo/CHANGELOG.md"
 write_postbump_state "$POSTBUMP_STATE" BUMP_TYPE=NONE
 OUT=$(run_subject postbump --state-file "$POSTBUMP_STATE" --implement-tmpdir "$SANDBOX/tmp")

